Feature/extras require, closes #165

caldwellst opened the pull request:
Monica! Putting this as draft because I have put lots of changes in here to make it run in the end, but I hope the functionality makes it through in some form.  So in general, my approach was to find a way to add extra requirements to the `setup.cfg` file. On the way, however, I have made some additional changes to make this happen. Here’s a general list of what I’ve done to guide your review.

1. By adding in the `sphinxcontrib.extras_require` section to `setup.cfg`, I started to explore the best way to reference files and install dev, doc, or whatever requirements. It seemed that incorporating all of his into `setup.cfg` was simpler than the mixed requirements files. I have tested and it seems to all work, and is cleaner, but maybe I’m missing some catch.
2. As part of this, I removed all requirements files, because it seems we can rely on pip to sort out dependencies however we install the package! Does that make sense?
3. Finally, I think that means `pip-tools` was no longer required for dev. My testing didn’t show any problems getting rid of it, but not sure of that.
4. Then I got ambitious. I discovered that you can automatically document the required extras using some sphinx references with `.. extras-require:`. However, for that to work it required that the sphinx base directory was only one level away from `setup.cfg`. Thus for the test to show what I’ve done, I put `docs-source` and `docs-build` separately within the main folder. But I know this might not be great in a final way, so just wanted to put that here for now for your draft review.
5. I did a quick check and I think GLOFAS is the only optional dependencies we should have now, but wasn't thorough, so we can add more before we merge if you like this.

caldwellst commented:
I think it's ready for review now! Took me a while to sort out exactly what approach to take for the optional dependencies, here were some specifics I dealt with:

1. Module-level imports of an optional dependency is impossible, since that generates errors any time you run `import aatoolbox` if the dependency is unavailable.
2. Rather than import in `__init__`, in which case I think I would've had to make `cdsapi` a property of the `Glofas` class, I just imported directly in the `_download()` method. However, I run the checks directly in `__init__`. Seems this might be a good model for any other optional dependencies going forward unless they are used across all methods.
3. For testing, I realized that it took me a while to ensure that the library doesn't error out if you don't have all dependencies. This is all based on the module-level implementation in Glofas rather than the checks. Thus I think each module with optional dependencies needs to specifically test that we can `import aatoolbox` without error if their dependencies aren't available, and that the correct error is generated if they are available.

Look forward to your review and hope the approach is logical!

turnerm commented on tests/datasources/test_glofas.py:
```diff
@@ -1,5 +1,7 @@
 """Test the GloFAS instantiation."""
+import sys
 from datetime import date
+from unittest.mock import patch
```
Better to use pytest for consistency with the other tests

turnerm commented on src/aatoolbox/utils/check_extra_imports.py:
```diff
@@ -0,0 +1,25 @@
+import importlib.util
+
+
+def _check_extra_imports(libraries: list, subpackage: str):
```
Should be public

turnerm commented on setup.cfg:
```diff
@@ -63,3 +61,25 @@ build-dir = docs/build
+    pytest-cov
+    pytest-mock
+    tox
+doc =
```
Are there not more sphinx packages required with this new setup?

caldwellst replied:
`extras_require` already comes with the sphinx extension installed, so that's activated in `conf.py`

turnerm reviewed (commented):
:see_no_evil: sorry this review is coming more than 2 months late, and after having done the whole rename so you're in for a horrible merge.

I think it looks really cool, and I like the idea to change the docs - I can't get them to compile though, are there some docs-related packages missing? I just want to see what Sphinx is doing exactly. It's too bad that this change means even more stuff (two docs dirs instead of one) in the main directory, but oh well, probably worth the price.

The only thing is that I would like to keep the pinned requirements files, these represent a single stable environment where we know everything works, and personally I think it's crucial to have in a project, even though it's not usually something used by end-users. Unfortunately it means putting things back into the `.in` files, unless you can somehow convince `pip-tools` to read the different sub-headings of `setup.cfg`.

caldwellst commented:
Alright! I've convinced `pip-tools` to use our new extras setup and recreated `requirements-dev.txt`. I didn't recreate the other files because now things like RTD can just install from the `setup.cfg`. I have also put things into a single doc folder with a `docs/_build` to store the build file and updated the contributing file. I think as you say it makes things cleaner in the repository and also isn't that confusing for the end user I think! It's also the default setup of the constructor from sphinx-build if you get a template (that's why it's setup that way in calendar) so I think is also "recommended" let's say.

I have merged in with develop and looks like it's ready to merge! Let me know what you think now.

turnerm reviewed (commented):
Sorry Seth, I know we have been iterating forever and it's my fault! A couple of more questions:
- I guess I realized that I don't actually understand what the `extras_require` package does for the documentation
- Relatedly, I think the installation options should be added to the documentation - maybe a summary on the main page, and a note in the GloFAS page?

Not to further derail everything, but I was looking at `flask` again and I noticed they are now using [`pip-compile-multi`](https://github.com/pallets/flask/blob/main/requirements/build.txt#L6) which solves the dependency issues we're always encountering. Their project also includes [optional depdencies](https://github.com/pallets/flask/blob/main/pyproject.toml#L40-L42).

caldwellst commented on docs/conf.py:
```diff
@@ -27,10 +27,12 @@
 # extensions coming with Sphinx (named 'sphinx.ext.*') or your custom
 # ones.
 extensions = [
+    "sphinx-prompt",
```
Have removed, wasn't necessary.

caldwellst commented on setup.cfg:
```diff
@@ -63,3 +61,25 @@ build-dir = docs/build
+full =
+    %(glofas)s
+test =
+    %(full)s
```
In this case, `full` is just all optional dependencies. I structured it this way so if we have data sources like `glofas` with optional dependencies, we can pass them to `test` and also to users who want a convenient way to install all optional decencies.

turnerm replied:
But `test` should be just what's needed to run the tests, as it was in `test.in`

caldwellst replied:
Yeah, but it needs the glofas dependencies to run the glofas tests.

caldwellst replied:
I tested it out, and even though those optional dependencies are mocked, the tests fail because the modules aren't installed.

caldwellst commented:
Absolutely no worries, I love this stuff and having fun going back through it. Apologies, I think I lost my mind at some point and removed some documentation and then got super busy and who knows what happened. I've added that stuff back. I still have a few changes to make to documentation, but wanted to first add this discussion point on requirements.


## Requirements files

Now, onto the debate of the requirements and how we address them. I have looked into `pip-compile-multi`, and here are my thoughts.

The benefits of `pip-compile-multi`:

- Ability to compile multiple requirements files together and address mismatched dependencies.
- Streamlined `pre-commit` since you simply need to have requirements files stored as `.in` files and it automatically takes care of the rest.

The cons of `pip-compile-multi`:

- Large number of requirements files instead of storing in one place together, see [flask](https://github.com/pallets/flask/tree/main/requirements) for example.
- It doesn't include the optional dependencies of `flask` automatically. For `flask`, the user is required to [manually install them](https://flask.palletsprojects.com/en/2.2.x/installation/#optional-dependencies). You would need to add individual `.in` files and then have a `full.in` file that would create the way to install the full package. 

The benefits of `extras_requires`:
- Ability to store all package requirements and extras in a single config (in this case `setup.cfg`), so basically streamlines the requirements files system
- With sphinx, can then easily reference the extra requirements for specific modules of our package, as I've done with glofas
<img width="959" alt="image" src="https://user-images.githubusercontent.com/8086547/226875763-43fbc25b-177a-4517-a7cc-38ed20f1ad68.png">
- Users can install the full package or optional extras with `pip install ocha-anticipy[full]` or `pip install ocha-anticipy[glofas]` without having to manually install dependencies, which I think would scale well.

The cons of `extras_requires`:
- You don't have the streamlined `pre-commit`, need to manually build all the `.txt` files we want.
- Lack of the `pip-compile-multi` dependency resolution across multiple `.txt` file requirements.

## My vote

Now, my understanding though is that `requirements-dev.txt` should have consolidated all dependencies necessary for production, testing, documentation, and development, since we have sourced the entirety of the package in there. Does `pip-compile-multi` actually provide any benefit on top of this? It might, as I'm not still totally clear on what it does.

However, I would vote for `extras_require` because it means we can streamline installation of optional extras to the user and add in the documentation with sphinx. It also maintains all the extras in the `setup.cfg` which in my mind is cleaner and easier to parse than having separate `.in` files to manage (although that's not actually that complex).

## Other question

My other question is, you asked to build the other requirements files, so I did. But I'm wondering, why are we doing that? Why don't we just need a development only requirements file that we use to develop from? Is it just in case we wanted to have a system only do documentation or only do testing we produce those? Since end users install from `pip`, why do we need all the different versions and not just the one dev version?

This is why I only built

turnerm reviewed (commented):
Okay, that image you shared DOES look super nice so I think that's convinced me. 

Regarding your question about the different requirements files: I think it's nice / good practice to have a requirements file that works for the situation you may need. So like if you only want to generate the docs, you only need to install a few packages.

This approach however breaks down with the `setup.cfg` implementation because all the base package requirements will be included in each requirements file, in addition to the extras. So maybe you're right that in this case it makes sense to have a single requirements file, which just has everything, and can be in the top level directory.

I also noticed that now if you make changes to `setup.cfg`, `pip-compile` recreates all the requirements files which takes a long time. So another reason to just have one. 

I think it looks great but the last thing is about the tests - do you think they can be made to not be dependent on extras being installed?

turnerm reviewed (approved):
Ah you're totally right about the tests. So I'm happy with it, thanks for all the cool new features!
